src/main/java/org/opensearch/searchrelevance/judgments/LlmJudgmentsProcessor.java

```diff
@@ -103,128 +110,111 @@ public void generateJudgmentRating(Map<String, Object> metadata, ActionListener<
-                judgmentForQuery.put("query", queryTextWithReference);
-                List<Map<String, String>> docIdRatings = docIdToScore.entrySet()
-                    .stream()
-                    .map(entry -> Map.of("docId", entry.getKey(), "rating", entry.getValue()))
-                    .collect(Collectors.toList());
-                judgmentForQuery.put("ratings", docIdRatings);
-                allJudgments.add(judgmentForQuery);
-
-                LOGGER.debug("Processed query: {} with {} ratings", queryTextWithReference, docIdRatings.size());
-            } catch (Exception e) {
-                LOGGER.error("Failed to process query: {}", queryTextWithReference, e);
-                if (!ignoreFailure) {
-                    throw new SearchRelevanceException("Failed to generate LLM judgments", e, RestStatus.INTERNAL_SERVER_ERROR);
-                }
-            }
-        }
-
-        LOGGER.info("Completed processing {} queries", queryTextWithReferences.size());
-        return allJudgments;
+        orchestrator.processQueriesConcurrently(
+            queryTextWithReferences,
+            (queryTextWithReference, queryListener) -> processQueryTextAsync(
+                modelId,
+                size,
+                tokenLimit,
+                contextFields,
+                searchConfigurations,
+                queryTextWithReference,
+                ignoreFailure,
+                queryListener
+            ),
+            ignoreFailure,
+            listener
+        );
     }
 
-    private Map<String, String> processQueryText(
+    private void processQueryTextAsync(
         String modelId,
         int size,
         int tokenLimit,
         List<String> contextFields,
         List<SearchConfiguration> searchConfigurations,
         String queryTextWithReference,
-        boolean ignoreFailure
+        boolean ignoreFailure,
+        ActionListener<Map<String, String>> listener
     ) {
-        Map<String, String> unionHits = new HashMap<>();
-        ConcurrentMap<String, String> docIdToScore = new ConcurrentHashMap<>();
-        Map<String, SearchHit> allHits = new HashMap<>();
+        JudgmentTaskContext context = new JudgmentTaskContext(
+            queryTextWithReference,
+            modelId,
+            contextFields,
+            searchConfigurations,
+            ignoreFailure,
+            listener
+        );
 
-        String queryText = queryTextWithReference.split(DELIMITER, 2)[0];
+        processQueryTextAsync(size, tokenLimit, context);
+    }
 
-        // Collect all hits
-        for (SearchConfiguration searchConfiguration : searchConfigurations) {
-            String index = searchConfiguration.index();
-            String query = searchConfiguration.query();
-            String searchPipeline = searchConfiguration.searchPipeline();
+    private void processQueryTextAsync(int size, int tokenLimit, JudgmentTaskContext context) {
+        Map<String, String> unionHits = new HashMap<>();
```

Comment:
this should be also a thread safe Map implementation, like ConcurrentHashMap for allHits, otherwise it's potentially unsafe in highly concurrent runtime